source/extensions/filters/http/transformation/inja_transformer.cc

```diff
@@ -65,22 +80,111 @@ Extractor::Extractor(const envoy::api::v2::filter::http::Extraction &extractor)
                     group_, extract_regex_.mark_count()));
   }
+
+  switch (mode_) {
+    case ExtractionApi::EXTRACT:
+      break;
+    case ExtractionApi::SINGLE_REPLACE:
+      if (!replacement_text_.has_value()) {
+        throw EnvoyException("SINGLE_REPLACE mode set but no replacement text provided");
+      }
+      break;
+    case ExtractionApi::REPLACE_ALL:
+      if (!replacement_text_.has_value()) {
+        throw EnvoyException("REPLACE_ALL mode set but no replacement text provided");
+      }
+      if (group_ != 0) {
+        throw EnvoyException("REPLACE_ALL mode set but subgroup is not 0");
+      }
+      break;
+    default:
+      throw EnvoyException("Unknown mode");
+  }
+}
+
+absl::optional<std::string>
+Extractor::getDynamicMetadataValue(Http::StreamFilterCallbacks &callbacks) const {
+  Upstream::ClusterInfoConstSharedPtr ci = callbacks.clusterInfo();
+  if (!ci) {
+    ENVOY_STREAM_LOG(debug, "No cluster info found for dynamic metadata", callbacks);
+    return absl::nullopt;
+  }
+
+  const envoy::config::core::v3::Metadata *cluster_metadata{};
+  cluster_metadata = &ci->metadata();
+  const ProtobufWkt::Value &value = Envoy::Config::Metadata::metadataValue(cluster_metadata, SoloHttpFilterNames::get().Transformation, dynamic_metadata_);
+
+  if (value.kind_case() == ProtobufWkt::Value::kStringValue && !value.string_value().empty()) {
+    return value.string_value();
+  } else if (value.string_value().empty()) {
+    ENVOY_STREAM_LOG(debug, "Dynamic metadata value not found at key: {}", callbacks, dynamic_metadata_);
+  }
+
+  return absl::nullopt; // Return empty if dynamic metadata not found or if value is empty
 }
 
 absl::string_view
 Extractor::extract(Http::StreamFilterCallbacks &callbacks,
                    const Http::RequestOrResponseHeaderMap &header_map,
                    GetBodyFunc &body) const {
-  if (body_) {
-    const std::string &string_body = body();
-    absl::string_view sv(string_body);
-    return extractValue(callbacks, sv);
-  } else {
-    const Http::HeaderMap::GetResult header_entries = getHeader(header_map, headername_);
-    if (header_entries.empty()) {
+  switch (source_) {
+    case ExtractionSource::BODY: {
+      const std::string &string_body = body();
+      return extractValue(callbacks, absl::string_view(string_body));
+    }
+    case ExtractionSource::HEADER: {
+      const Http::HeaderMap::GetResult header_entries = getHeader(header_map, headername_);
+      if (header_entries.empty()) {
+        return "";
+      }
+      return extractValue(callbacks, header_entries[0]->value().getStringView());
+    }
+    case ExtractionSource::DYNAMIC_METADATA: {
+      auto value = getDynamicMetadataValue(callbacks);
+      if (value) {
+        return extractValue(callbacks, *value);
+      }
+      return "";
+    }
+    default:
+      throw EnvoyException("Unsupported extraction source");
+  }
+}
+
+std::string
+Extractor::replace(Http::StreamFilterCallbacks &callbacks,
```

Comment:
Look into whether it makes sense to extract shared code b/w this and `Extractor::extract`

Reply:
Consider renaming replace to destructive extraction/extract destructive, etc.

Reply:
Updated here: https://github.com/solo-io/envoy-gloo/pull/309/commits/741ea24a34a8ecfce5c8b5a39f4bd13ab69bf71e